importccl: support AUTO_INCREMENT in mysqldump

dt opened the pull request:
This adds support for automatically creating a sequence and using it in the default value of AUTO_INCREMENT mysql columns.
This functions similarly to the pgdump functionality with a couple differences:
  - the mysqldump does not explicitly create a sequence or reference it, so we must make one and use that
  - the mysqldump format includes the next value in the table definition, so we need to capture it when we read the table,
    and then use it to set the sequence later.

Fixes #28711.

Release note: none.

dt committed:
importccl: support setting inital sequence values on success

This adds support to the IMPORT job tracking to support setting starting sequence values based on values determined during job initialization.
This differs from the current handling of these sequences in pgdump, where the 'setval' call is seen hand handled during the data import, and
thus simply emits the correctly formed KV along with the other table data.

Release note: none.

dt committed:
importccl: use MakeSimpleTableDescriptor in mysql import

Previously we handled each mysql AST def by adding elements individual to a hand-made TableDescriptor.
This however meant we often missed common functionality in MakeSimpleTableDescriptor and the various
SQL helpers for creating well-formed TableDescriptors will all the right book-keeping in place.

This switches the mysqldump table conversion to convert a MySQL AST CREATE TABLE to a Cockroach AST
one, then hand that AST to the usual MakeSimpleTableDescriptor and let it figure it out.

Release note: none.

dt committed:
importccl: support AUTO_INCREMENT in mysqldump

This adds support for automatically creating a sequence and using it in the default value of AUTO_INCREMENT mysql columns.
This functions similarly to the pgdump functionality with a couple differences:
  - the mysqldump does not explicitly create a sequence or reference it, so we must make one and use that
  - the mysqldump format includes the next value in the table definition, so we need to capture it when we read the table,
    and then use it to set the sequence later.

Release note: none.

knz reviewed (commented):
I am surprised by the lack of release notes.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, 2 of 2 files at r2, 12 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28761#-:-LK7Kg-YBDMFnroD-yEf:b-us4ab)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

dt commented:
@knz all of the MYSQLDUMP support is new, so no changes to user-facing released functionality.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -288,15 +289,41 @@ func mysqlTableToCockroach(
+			priv := sqlbase.NewDefaultPrivilegeDescriptor()
+			opts := tree.SequenceOptions{{Name: tree.SeqOptStart, IntVal: &i64}}
+			seqName = name + "_auto_inc"
+			desc, err := sql.MakeSequenceTableDesc(seqName, opts, parentID, id, time, priv, nil)
```
add comment for the nil

dt replied:
done.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -306,6 +333,16 @@ func mysqlTableToCockroach(
+			if seqName == "" {
+				return nil, nil, errors.Errorf("column %q specifies AUTO_INCREMENT but table options did not include it", def.Name)
+			}
+			expr, err := parser.ParseExpr(fmt.Sprintf("nextval('%s':::STRING)", seqName))
```
is it possible for seqName to contain a single quote? Seems possible if the table name contains a single quote?

dt replied:
Added another commit that uses [`CompliantName`](https://github.com/vitessio/vitess/blob/master/go/vt/sqlparser/ast.go#L3597) everywhere, which replaces any non-letter, and any non-digit or leading digit with `_` as far as I can tell.

knz commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -294,55 +294,42 @@ func mysqlTableToCockroach(
 
-	colNames := make(map[string]string)
+	stmt := &tree.CreateTable{
+		Table: tree.NormalizableTableName{tree.NewTableName("", tree.Name(name))},
```
yes correct

dt replied:
Done.

dt committed:
importccl: use normalized, complaint names in mysqldump import

This uses the 'compliant' names for all mysql identifiers -- lower-case with any non-letters and non-digits replaced with '_'.

Release note: none.

dt commented:
This is ready for another look -- hoping to merge it by tomorrow.

dt commented:
bors r+

merging on knz's lgtm to make branch cutting day, but if there are any bugs I missed happy to fix in follow up
